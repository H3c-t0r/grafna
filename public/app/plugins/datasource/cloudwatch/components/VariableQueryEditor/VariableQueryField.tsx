import { SelectableValue } from '@grafana/data';
import { InlineField, Select } from '@grafana/ui';
import React, { FC } from 'react';

const LABEL_WIDTH = 20;

interface VariableQueryFieldProps {
  onChange: (value: string) => void;
  options: SelectableValue[];
  value: string | null;
  label: string;
  allowCustomValue?: boolean;
  isLoading?: boolean;
  inputId?: string;
}

export const VariableQueryField: FC<VariableQueryFieldProps> = ({
  label,
  onChange,
  value,
  options,
  allowCustomValue = false,
  isLoading = false,
}) => {
  return (
    <InlineField label={label} labelWidth={LABEL_WIDTH} htmlFor={'inline-field'}>
      <Select
        menuShouldPortal
        aria-label={label}
        width={25}
        allowCustomValue={allowCustomValue}
        value={value}
        onChange={({ value }) => onChange(value!)}
        options={options}
        isLoading={isLoading}
        inputId="inline-field"
      />
    </InlineField>
  );
};
