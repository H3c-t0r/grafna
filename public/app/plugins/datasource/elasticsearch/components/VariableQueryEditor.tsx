import React, { PureComponent } from 'react';
import { VariableQueryProps } from 'app/types/plugins';
import { Select, FormLabel, SelectOptionItem } from '@grafana/ui';
import { MetricFindQueryTypes } from '../types';
import { FieldsQueryForm } from './FieldsQueryForm';
import { TermsQueryForm } from './TermsQueryForm';

export interface VariableQueryState {
  find: string;
  fields?: SelectOptionItem[];
  initialQuery?: any;
}

const queryTypes: SelectOptionItem[] = [
  { value: MetricFindQueryTypes.Fields, label: 'Fields' },
  { value: MetricFindQueryTypes.Terms, label: 'Terms' },
];

const defaultState: VariableQueryState = {
  find: queryTypes[0].value,
  fields: [],
};

export class ElasticVariableQueryEditor extends PureComponent<VariableQueryProps, VariableQueryState> {
  constructor(props: VariableQueryProps) {
    super(props);

    let query: any = {};
    if (this.props.query && this.props.query.length > 0) {
      query = JSON.parse(this.props.query);
    }

    query.initialQuery = { ...query };
    this.state = {
      ...defaultState,
      ...query,
    };
  }

  async componentDidMount() {
    const { templateSrv } = this.props;
    const fieldOptions = await this.getFields();
    const templateVariables = templateSrv.variables.map(v => ({
      label: `$${v.name}`,
      value: `$${v.name}`,
      description: v.current && v.current.value ? `Current value: ${v.current.value}` : v.label || v.name,
    }));
    const fields = [...templateVariables, ...fieldOptions];

    this.setState({
      fields,
    });
  }

  async getFields() {
    const { datasource } = this.props;
    return datasource.getFields({}).then((result: string[]) => {
      return result.map((field: any) => {
        return {
          value: field.text,
          label: field.text,
          description: field.type,
        };
      });
    });
  }

  onQueryTypeChange = (item: SelectOptionItem) => {
    this.setState({
      find: item.value,
    });
  };

  onQueryChange = (query: any, definition: string) => {
    const jsonQuery = JSON.stringify(query);
    this.props.onChange(jsonQuery, definition);
  };

  renderQueryType() {
    const { find, initialQuery, fields } = this.state;

    switch (find) {
      case MetricFindQueryTypes.Fields:
        return <FieldsQueryForm query={initialQuery} onChange={this.onQueryChange} />;
      case MetricFindQueryTypes.Terms:
        return <TermsQueryForm query={initialQuery} fields={fields} onChange={this.onQueryChange} />;
    }

    return '';
  }

  render() {
    const { find } = this.state;

    return (
      <>
        <div className="form-field">
          <FormLabel className="query-keyword">Query Type</FormLabel>
          <Select
            placeholder="Choose type"
            isSearchable={false}
            options={queryTypes}
            value={queryTypes.find(o => o.value === find)}
            onChange={this.onQueryTypeChange}
            width={11}
          />
        </div>
        {this.renderQueryType()}
      </>
    );
  }
}
