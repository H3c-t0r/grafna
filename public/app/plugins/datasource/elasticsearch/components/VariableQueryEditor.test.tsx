import React from 'react';
import { shallow } from 'enzyme';
import { ElasticVariableQueryEditor } from './VariableQueryEditor';

const setup = async (propOverrides?: object) => {
  const props = {
    onChange: () => {},
    query: '',
    datasource: {
      getFields: async () => Promise.resolve([{ text: 'field', type: 'number' }]),
    },
    templateSrv: { variables: [{ name: 'test' }] },
    ...propOverrides,
  };

  const wrapper = await shallow(<ElasticVariableQueryEditor {...props} />);
  const instance = wrapper.instance() as ElasticVariableQueryEditor;

  return {
    wrapper,
    instance,
  };
};

describe('ElasticVariableQueryEditor', () => {
  it('should render component', async () => {
    const { wrapper } = await setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('when loading component should initialize state', async () => {
    const { instance } = await setup();
    await instance.componentDidMount();
    const { find, fields, initialQuery } = instance.state;
    expect(find).toBe('fields');
    expect(fields).toEqual([{ value: 'field', label: 'field', description: 'number' }]);
    expect(initialQuery).toEqual({});
  });

  it('when loading component with existing fields query should initialize state', async () => {
    const { instance } = await setup({
      query: JSON.stringify({ find: 'fields', type: 'keyword' }),
    });
    await instance.componentDidMount();
    const { find, initialQuery } = instance.state;
    expect(find).toBe('fields');
    expect(initialQuery).toEqual({
      find: 'fields',
      type: 'keyword',
    });
  });

  it('when loading component with existing terms query should initialize state', async () => {
    const { instance } = await setup({
      query: JSON.stringify({ find: 'terms', field: 'field' }),
    });
    await instance.componentDidMount();
    const { find, initialQuery } = instance.state;
    expect(find).toBe('terms');
    expect(initialQuery).toEqual({
      find: 'terms',
      field: 'field',
    });
  });

  it('should convert query to json on change', async () => {
    const props = {
      onChange: jest.fn(),
    };
    const { instance } = await setup(props);
    const query = { find: 'fields', type: 'keyword' };
    instance.onQueryChange(query, 'def');
    expect(props.onChange.mock.calls).toHaveLength(1);
    expect(props.onChange.mock.calls[0][0]).toBe(JSON.stringify(query));
    expect(props.onChange.mock.calls[0][1]).toBe('def');
  });
});
