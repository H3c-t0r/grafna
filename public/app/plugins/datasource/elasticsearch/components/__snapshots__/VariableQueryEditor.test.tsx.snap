// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticVariableQueryEditor should render component 1`] = `
<Fragment>
  <div
    className="form-field"
  >
    <Component
      className="query-keyword"
    >
      Query Type
    </Component>
    <Select
      autoFocus={false}
      backspaceRemovesValue={true}
      className=""
      components={
        Object {
          "Group": [Function],
          "IndicatorsContainer": [Function],
          "MenuList": [Function],
          "Option": [Function],
          "SingleValue": [Function],
        }
      }
      isClearable={false}
      isDisabled={false}
      isLoading={false}
      isMulti={false}
      isSearchable={false}
      maxMenuHeight={300}
      menuIsOpen={false}
      onChange={[Function]}
      openMenuOnFocus={false}
      options={
        Array [
          Object {
            "label": "Fields",
            "value": "fields",
          },
          Object {
            "label": "Terms",
            "value": "terms",
          },
        ]
      }
      placeholder="Choose type"
      value={
        Object {
          "label": "Fields",
          "value": "fields",
        }
      }
      width={11}
    />
  </div>
  <FieldsQueryForm
    onChange={[Function]}
    query={Object {}}
  />
</Fragment>
`;
